### PR TITLE
backend: Adding decorator for compass metrics. Adding create and delete index …

### PR DESCRIPTION
**Pull request by ankush-cohere**
Thank you for contributing to the Cohere Toolkit!

- [x] **PR title**: "backend: Adding decorator for compass metrics. Adding create and delete index method in compass tool"

- [x] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** Adding decorator to cover metrics around compass
    - **Issue:** [SEA-1356](https://linear.app/cohereai/issue/SEA-1356/send-message-to-reporting-api-from-compass)
    - **Dependencies:** No

- [ ] **Add tests and docs**: Please include testing and documentation for your changes
- [ ] **Lint and test**: Run `make lint` and `make run-tests`

**AI Description**

The PR makes changes to the model deployment and metrics collection code. 

## Summary
The changes include:
- Updates to the import statements in `azure.py, `bedrock.py, `cohere_platform.py, `sagemaker.py, and `metrics.py.
- Addition of the `collect_compass_metrics` function in `metrics.py.
- Updates to the function decorators in `azure.py, `bedrock.py, `cohere_platform.py, `sagemaker.py, and `metrics.py.
- Updates to the `invoke_chat` and `invoke_chat_stream` functions in `azure.py, `bedrock.py, `cohere_platform.py, and `sagemaker.py.
- Updates to the `run_loop` function in `metrics.py.
- Updates to the `wrap_and_log_data` function in `metrics.py.
- Updates to the `collect_metrics_chat`, `collect_metrics_chat_stream`, and `collect_metrics_rerank` functions in `metrics.py.
- Updates to the `call` function in `compass.py.
- Addition of the `_create_index` and `_delete_index` functions in `compass.py.

## Details
- The import statements have been updated to remove redundant imports and add necessary ones.
- The `collect_compass_metrics` function is added to `metrics.py` to collect metrics for the Compass tool.
- The function decorators have been updated to include the endpoint and method as parameters.
- The `invoke_chat` and `invoke_chat_stream` functions now include the endpoint and method as parameters in the function decorators.
- The `run_loop` function now includes additional parameters for the endpoint name, method, and model.
- The `wrap_and_log_data` function now includes the endpoint and method as parameters in the `curl` command.
- The `collect_metrics_chat`, `collect_metrics_chat_stream`, and `collect_metrics_rerank` functions have been updated to include the endpoint and method as parameters.
- The `call` function in `compass.py` now returns an empty list instead of `None` in certain cases.
- The `_create_index` and `_delete_index` functions have been added to `compass.py` to create and delete indexes, respectively.

### CODE REVIEW

File: src/backend/services/metrics.py [hunk 1]
```diff
@@ -297,17 +298,54 @@ def wrapper(self, query: str, documents: Dict[str, Any], **kwargs: Any) -> Any:
...
+                metrics_data.duration_ms = time.perf_counter() - start_time
+                run_loop(metrics_data)
+                return response_dict
+
+        return wrapper
+
+    return report_metrics
+
+
 def initialize_sdk_metrics_data(
```
Comment (lusmoura) [inline src/backend/services/metrics.py:327]: Can you unify these two functions? You can change the existing decorators to also receive the name and method!

File: src/backend/services/metrics.py [hunk 2]
```diff
@@ -224,99 +223,155 @@ def run_loop(metrics_data: MetricsData) -> None:
...
-        except Exception as e:
-            metrics_data = handle_error(metrics_data, e)
-            raise e
-        finally:
-            metrics_data.duration_ms = time.perf_counter() - start_time
-            run_loop(metrics_data)
-            return response_dict
 
-    return wrapper
+# def initialize_sdk_metrics_data(
```
Comment (lusmoura) [inline src/backend/services/metrics.py:353]: Do we want to keep this?

Comment (ankush-cohere) [reply src/backend/services/metrics.py:353]: Nope. Will remove

**Comment by ankush-cohere** [comment]
Closing as we have diverged away from decorator solution. 
cc: @tanzimCohere @lusmoura @benrules3

**Pull request closed without merging**